### PR TITLE
Switch MSBuildWorkspace to use UTF-8 without BOM

### LINKED ISSUE #1146: Workspaces should not be outputing an explicit BOM
Consider the code inside `Workspace.ApplyDocumentTextChange`:

- https://github.com/dotnet/roslyn/blob/master/src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildWorkspace.cs#L972

This code is using an explicit UTF8 BOM in the case an encoding can't be detected.  This seems like a bug.  Shouldn't the default be a no-BOM UTF8 encoding?

### PR DESCRIPTION
**Pull request by Maxwe11**
Fixes #1146

### CODE REVIEW

**Comment by Pilchie** [comment]
LGTM.  @mattwar, thoughts?

File: src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildWorkspace.cs [hunk 1]
```diff
@@ -997,7 +997,7 @@ protected override void ApplyDocumentTextChanged(DocumentId documentId, SourceTe
             {
                 Encoding encoding = DetermineEncoding(text, document);
 
-                this.SaveDocumentText(documentId, document.FilePath, text, encoding ?? Encoding.UTF8);
+                this.SaveDocumentText(documentId, document.FilePath, text, encoding ?? new UTF8Encoding(false));
```
Comment (KevinH-MS) [inline src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildWorkspace.cs]: Consider "encoderShouldEmitUTF8Identifier: false" (use named argument)

**Comment by KevinH-MS** [comment]
:+1:

File: src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildWorkspace.cs (hunk 1 above)
Comment (jaredpar) [inline src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildWorkspace.cs]: :+1:

**Comment by jaredpar** [comment]
Is there a particular reason why we have a different default encoding here than in `EncodedStringText`.  Seems we prefer code page 1252 there and UTF8 here.  True it's default for read and default for write.  Would like to understand the reasoning though.

**Comment by KevinH-MS** [comment]
I'm not looking at the ```EncodedStringText``` code, but my mental model of how it should work is that reading should try UTF8 and fall back to 1252 (if there are any characters that cannot be decoded).  Writing should prefer UTF8.

**Comment by jaredpar** [comment]
@KevinH-MS okay, that makes sense.

**Comment by jaredpar** [comment]
LGTM

File: src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildWorkspace.cs (hunk 1 above)
Comment (Maxwe11) [inline src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildWorkspace.cs]: will fix

**Comment by Maxwe11** [comment]
updated

**Comment by mattwar** [comment]
LGTM2

**Comment by pharring** [comment]
@Maxwe11 Please add a unit test so we can lock in this behavior.

**Comment by jasonwilliams200OK** [comment]
:+1: 
Slightly related: https://github.com/dotnet/coreclr/issues/1012
Love this line:

> The world has spoken, and UTF8 is the standard, we should have first-class support for it both for P/Invoke signatures as well as the various helper methods in Marshal.

With ASCII only characters, a UTF-16 encoded file is roughly twice the size compared to one encoded with UTF-8.

**Comment by Pilchie** [comment]
Oops, merged without seeing @pharring's comment.  A test would definitely be good here.

**Pull request merged**